staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset_pruner.go

```diff
@@ -16,15 +16,132 @@ limitations under the License.
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/client-go/dynamic"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+)
 
 type applySetPruner struct {
+	dynamicClient dynamic.Interface
+
+	visitedUids sets.Set[types.UID]
+
+	cascadingStrategy metav1.DeletionPropagation
+	dryRunStrategy    cmdutil.DryRunStrategy
+	gracePeriod       int
+
+	printer printers.ResourcePrinter
+
+	out io.Writer
+}
+
+func newApplySetPruner(o *ApplyOptions) (*applySetPruner, error) {
+	printer, err := o.ToPrinter("pruned")
+	if err != nil {
+		return nil, err
+	}
+
+	return &applySetPruner{
+		dynamicClient: o.DynamicClient,
+
+		visitedUids: o.VisitedUids,
+
+		cascadingStrategy: o.DeleteOptions.CascadingStrategy,
+		dryRunStrategy:    o.DryRunStrategy,
+		gracePeriod:       o.DeleteOptions.GracePeriod,
+
+		printer: printer,
+
+		out: o.Out,
+	}, nil
+}
+
+func formatGVKForHumans(gvk schema.GroupVersionKind) string {
+	gk := gvk.GroupKind()
+	return gk.Kind
 }
 
-func newApplySetPruner(_ *ApplyOptions) *applySetPruner {
-	return &applySetPruner{}
+func formatNamespacedNameForHumans(id types.NamespacedName) string {
+	if id.Namespace == "" {
+		return id.Name
+	}
+	return id.Namespace + "/" + id.Name
+}
+
+func (p *applySetPruner) pruneAll(ctx context.Context, applyset *ApplySet) error {
+	// TODO: Split into discovery and deletion, run discovery in parallel (and maybe in consistent order or in parallel?)
+	for _, restMapping := range applyset.AllPrunableResources() {
+		switch restMapping.Scope.Name() {
+		case meta.RESTScopeNameNamespace:
+			for _, namespace := range applyset.AllPrunableNamespaces() {
+				if err := p.prune(ctx, namespace, restMapping, applyset); err != nil {
+					return fmt.Errorf("pruning %v objects: %w", formatGVKForHumans(restMapping.GroupVersionKind), err)
+				}
+			}
+
+		case meta.RESTScopeNameRoot:
+			if err := p.prune(ctx, metav1.NamespaceNone, restMapping, applyset); err != nil {
+				return fmt.Errorf("pruning %v objects: %w", formatGVKForHumans(restMapping.GroupVersionKind), err)
+			}
+
+		default:
+			return fmt.Errorf("unhandled scope %q", restMapping.Scope.Name())
+		}
+	}
+
+	return nil
+}
+
+func (p *applySetPruner) prune(ctx context.Context, namespace string, mapping *meta.RESTMapping, applyset *ApplySet) error {
+	applysetLabelSelector := applyset.LabelSelectorForMembers()
+
+	opt := metav1.ListOptions{
+		LabelSelector: applysetLabelSelector,
+	}
+
+	objects, err := p.dynamicClient.Resource(mapping.Resource).Namespace(namespace).List(ctx, opt)
+	if err != nil {
+		return err
+	}
+
+	for i := range objects.Items {
+		obj := &objects.Items[i]
+
+		// Unlike prunev1, we prune resources whether they were created with
+		// client-side-apply, server-side-apply or indeed any method.
+		// prunev1 checked the last-applied annotation as a safety check against accidental deletion,
+		// but our label should be at least as good a check, if not better.
```

Comment:
Nite: I know you added these lines in response to my comment, but I'd suggest they aren't necessary. Most people don't know how prune v1 worked, and the CSA annotation just isn't relevant to us at all anymore (would produce incorrectness, not improve safety).

Reply:
Removed!